iree/compiler/Dialect/Flow/IR/PartitionableLoopsInterface.cpp

```diff
@@ -0,0 +1,197 @@
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Flow/IR/PartitionableLoopsInterface.h"
+
+#include "iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.h"
+#include "iree-dialects/Dialect/LinalgExt/IR/TiledOpInterface.h"
+#include "llvm/ADT/SmallVector.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+
+using namespace llvm;
+
+#include "iree/compiler/Dialect/Flow/IR/PartitionableLoopsInterface.cpp.inc"  // IWYU pragma: export
+
+/// Filters out dimensions in `parallelLoops` that have unit range in
+/// `loopRanges`.
+static SmallVector<unsigned> pruneUnitTripParallelLoops(
+    ArrayRef<unsigned> parallelLoops, ArrayRef<int64_t> loopRanges) {
+  return llvm::to_vector(llvm::make_filter_range(
+      parallelLoops,
+      [&loopRanges](unsigned loopDim) { return loopRanges[loopDim] != 1; }));
+}
+
+/// Returns the partitionable loops for all Linalg ops.
+SmallVector<unsigned> getPartitionableLoopsImpl(
+    mlir::linalg::LinalgOp linalgOp, unsigned maxNumPartitionedLoops) {
+  SmallVector<unsigned> parallelLoops;
+  linalgOp.getParallelDims(parallelLoops);
+  // Get the static loop ranges.
+  Optional<SmallVector<int64_t, 4>> staticLoopRanges =
+      linalgOp.getStaticLoopRanges();
+  if (staticLoopRanges) {
+    parallelLoops =
+        pruneUnitTripParallelLoops(parallelLoops, *staticLoopRanges);
+  }
+  // TODO(ravishankarm): For now the outer parallel loops are dropped. This is
+  // a pragmatic choice for now but might need to be revisited.
+  if (parallelLoops.size() > maxNumPartitionedLoops) {
+    parallelLoops = llvm::to_vector(
+        ArrayRef<unsigned>(parallelLoops).take_back(maxNumPartitionedLoops));
+  }
+  return parallelLoops;
+}
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+/// External model implementation for all LinalgOps.
+struct LinalgOpPartitionableLoops
+    : public PartitionableLoopsInterface::ExternalModel<
+          LinalgOpPartitionableLoops, linalg::LinalgOp> {
+  unsigned getNumLoops(Operation *op) const {
+    auto linalgOp = cast<linalg::LinalgOp>(op);
+    return linalgOp.getNumLoops();
+  }
+
+  SmallVector<unsigned> getPartitionableLoops(
+      Operation *op, unsigned maxNumPartitionedLoops) const {
+    auto linalgOp = cast<linalg::LinalgOp>(op);
+    return getPartitionableLoopsImpl(linalgOp, maxNumPartitionedLoops);
+  }
+};
+
+/// External model implementation for linalg::Mmt4DOp;
+struct Mmt4DOpPartitionableLoops
+    : public PartitionableLoopsInterface::ExternalModel<
+          Mmt4DOpPartitionableLoops, linalg::Mmt4DOp> {
+  unsigned getNumLoops(Operation *op) const {
+    auto linalgOp = cast<linalg::LinalgOp>(op);
+    return linalgOp.getNumLoops();
+  }
+
+  SmallVector<unsigned> getPartitionableLoops(
+      Operation *op, unsigned maxNumPartitionedLoops) const {
+    return {0, 1};
```

Comment:
This is not considering `maxNumPartitionedLoops` so it could break the contract as defined in the doc. Should also check and make sure not returning more than `maxNumPartitionedLoops`.